k9mail-library/src/main/java/com/fsck/k9/mail/Flag.java

```diff
@@ -19,40 +41,183 @@
      */
-    X_DOWNLOADED_PARTIAL,
+    public static final Flag X_DOWNLOADED_PARTIAL = new Flag("X_DOWNLOADED_PARTIAL");
 
     /**
      * Indicates that the copy of a message to the Sent folder has started.
      */
-    X_REMOTE_COPY_STARTED,
+    public static final Flag X_REMOTE_COPY_STARTED = new Flag("X_REMOTE_COPY_STARTED");
 
     /**
      * Indicates that all headers of the message have been stored in the
      * database. If this is false, additional headers might be retrieved from
      * the server (if the message is still there).
      */
-    X_GOT_ALL_HEADERS,
+    public static final Flag X_GOT_ALL_HEADERS = new Flag("X_GOT_ALL_HEADERS");
+
+    /*
+     * Predefined Prefixes
+     */
+    private static final String USER_PREFIX = "USER_";
+
+    // when internal name = name we refer to it as just name
+    private final String mName;             // for use towards third party  ex. "\\Deleted"
+    private final String mInternalName;     // for internal use in database,...   ex. "DELETED"
+    protected final boolean mCustom;
+
+    /**
+     * When a Flag is created dynamically we know it's a custom flag.
+     *
+     * @param mName Internal name of the flag.
+     * @return Newly created Flag object.
+     */
+    public static Flag createFlag(String name) {
+        Flag tmpFlag = new Flag(USER_PREFIX + name, name, true);
+        return tmpFlag;
+    }
+
+    private Flag(String name) {
+        this(name, name);
+    }
+
+    /**
+     * Create a new Flag. This doesn't create a custom flag, it's used to define
+     * the predefined flags.
+     *
+     * @param internalName Name for internal use in database,...   ex. "DELETED"
+     * @param name Name for use towards third party ( ex. "\\Deleted" )
+     */
+    private Flag(String internalName, String name) {
+        this(internalName, name, false);
+    }
+
+    private Flag(String internalName, String name, boolean isCustom) {
+        this.mName = name;
+        this.mCustom = isCustom;
+        this.mInternalName = internalName;
+    }
+
+    /**
+     * Returns the predefined static flag object if any. Otherwise a new
+     * custom flag is created and returned.
+     *
+     * When the name starts with the USER_PREFIX we don't add it again to
+     * create a new one. This is the case for example when this is called with
+     * strings retrieved from the database.
+     *
+     * IMPORTANT remember the name of the field of predefined flags must equal the
+     * internal name!
+     *
+     * @param mName Name of Flag wanted.
+     * @return  Predefined Flag object if any otherwise new custom Flag.
+     * @throws IllegalArgumentException Thrown when the field is not accessible.
+     */
+    public static Flag valueOf(String internalName) throws IllegalArgumentException {
+        try {
+            // the argument being 'null' implies we look for a static field
+            return (Flag)(Flag.class.getField(internalName).get(null));
```

Comment:
Rather than reflection here, I would extract the internal flags listed in valueOfByRealName into a method returning a set of Strings and see if it contains internalName.

You can ensure the new method is kept in sync with the actual fields in a unit test.The unit test can use reflection to enumerate Flag's fields without concern about performance.